viamonvif/discovery.go

```diff
@@ -1,241 +1,120 @@
+	GetStreamURI(profile onvif.Profile, creds device.Credentials) (*url.URL, error)
 }
 
-// callAndParse sends a request to an ONVIF device, decoding and mutating the
-// provided response struct with the result. Returns an error on failure.
-func callAndParse(logger logging.Logger, deviceInstance OnvifDevice, request interface{}, response any) error {
-	resp, err := deviceInstance.CallMethod(request)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
+// DiscoverCameras performs WS-Discovery
+// then uses ONVIF queries to get available RTSP addresses and supplementary info.
+func DiscoverCameras(creds []device.Credentials, manualXAddrs []*url.URL, logger logging.Logger) (*CameraInfoList, error) {
+	var discoveredCameras []CameraInfo
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
 
-	body, err := io.ReadAll(resp.Body)
+	logger.Debug("WS-Discovery start")
+	ifaces, err := net.Interfaces()
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("failed to retrieve network interfaces: %w", err)
 	}
 
-	logger.Debugf("raw response: %v", string(body))
-
-	// Reset the response body reader after logging
-	resp.Body = io.NopCloser(bytes.NewReader(body))
-
-	return xml.NewDecoder(resp.Body).Decode(response)
-}
-
-// getProfilesResponse is the schema the GetProfiles response is formatted in.
-type getProfilesResponse struct {
-	XMLName xml.Name `xml:"Envelope"`
-	Body    struct {
-		GetProfilesResponse struct {
-			Profiles []onvifxsd.Profile `xml:"Profiles"`
-		} `xml:"GetProfilesResponse"`
-	} `xml:"Body"`
-}
-
-// getStreamURIResponse is the schema the GetStreamUri response is formatted in.
-type getStreamURIResponse struct {
-	XMLName xml.Name `xml:"Envelope"`
-	Body    struct {
-		GetStreamURIResponse struct {
-			MediaURI onvifxsd.MediaUri `xml:"MediaUri"`
-		} `xml:"GetStreamUriResponse"`
-	} `xml:"Body"`
-}
+	discovered := map[string]*url.URL{}
+	for _, xaddr := range manualXAddrs {
+		discovered[xaddr.Host] = xaddr
+	}
+	for _, iface := range ifaces {
+		xaddrs, err := WSDiscovery(ctx, logger, iface)
+		if err != nil {
+			logger.Debugf("WS-Discovery skipping interface %s: due to error from SendProbe: %w", iface.Name, err)
+			continue
+		}
+		for _, xaddr := range xaddrs {
+			discovered[xaddr.Host] = xaddr
+		}
+	}
+	logger.Debug("WS-Discovery complete")
 
-// getDeviceInformationResponse is the schema the GetDeviceInformation response is formatted in.
-type getDeviceInformationResponse struct {
-	XMLName xml.Name `xml:"Envelope"`
-	Body    struct {
-		GetDeviceInformationResponse struct {
-			Manufacturer string `xml:"Manufacturer"`
-			Model        string `xml:"Model"`
-			SerialNumber string `xml:"SerialNumber"`
-		} `xml:"GetDeviceInformationResponse"`
-	} `xml:"Body"`
+	for _, xaddr := range discovered {
+		cameraInfo, err := DiscoverCamerasOnXAddr(ctx, xaddr, creds, logger)
+		if err != nil {
+			logger.Warnf("failed to connect to ONVIF device %v", err)
```

Comment:
TODO: We should probably not log a warning if it is using different credentials, given it might be that this will always log warnings if there are cameras with different credentials on the same network responding to WS-Discovery messages.